beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+	storedCount := 0
+
+	log.Error("AAAA")
+
+	for {
+		dataColumnSidecars := pullChan(inputDataColumnSidecars)
+		if len(dataColumnSidecars) == 0 {
+			break
+		}
+
+		for _, dataColumnSidecar := range dataColumnSidecars {
+			// Extract the data column index.
+			dataColumnIndex := dataColumnSidecar.ColumnIndex
+
+			// Skip if the data column is already stored.
+			if indices[dataColumnIndex] >= limit {
+				continue
+			}
+
+			// Alter the indices to mark the first data column sidecar as saved.
+			// savedCount can safely be cast to uint8 since it is less than limit.
+			indices[dataColumnIndex] = limit + uint8(storedCount)
+
+			// Increment the count of the saved SSZ encoded data column sidecar.
+			storedCount++
+
+			// SSZ encode the first data column sidecar.
+			sszEncodedDataColumnSidecar, err := dataColumnSidecar.MarshalSSZ()
+			if err != nil {
+				return errors.Wrap(err, "data column sidecar marshal SSZ")
+			}
+
+			// Check if the size of the SSZ encoded data column sidecar is correct.
+			if sszEncodedDataColumnSidecarRefSize != 0 && len(sszEncodedDataColumnSidecar) != sszEncodedDataColumnSidecarRefSize {
+				return ErrWrongSszEncodedDataColumnSidecarSize
+			}
+
+			// Set the SSZ encoded data column sidecar reference size.
+			sszEncodedDataColumnSidecarRefSize = len(sszEncodedDataColumnSidecar)
+
+			// Append the first SSZ encoded data column sidecar to the SSZ encoded data column sidecars.
+			sszEncodedDataColumnSidecars = append(sszEncodedDataColumnSidecars, sszEncodedDataColumnSidecar...)
+		}
+	}
+
+	if storedCount == 0 {
+		// Nothing to save.
+		return nil
+	}
+
+	// Create the data column sidecars file.
+	dir := filepath.Dir(filePath)
+	if err := dcs.fs.MkdirAll(dir, directoryPermissions()); err != nil {
+		return errors.Wrapf(err, "mkdir all")
+	}
+
+	file, err := dcs.fs.Create(filePath)
+	if err != nil {
+		return errors.Wrap(err, "data column sidecars file path create")
+	}
+
+	defer func() {
+		closeErr := file.Close()
+
+		// Overwrite the existing error only if it is nil, since the close error is less important.
+		if closeErr != nil && err == nil {
+			err = closeErr
+		}
+	}()
+
+	// Encode the SSZ encoded data column sidecar size.
+	var encodedSszEncodedDataColumnSidecarSize [encodedSszEncodedDataColumnSidecarSizeSize]byte
+	binary.BigEndian.PutUint32(encodedSszEncodedDataColumnSidecarSize[:], uint32(sszEncodedDataColumnSidecarRefSize))
+
+	// Concatenate the version, the data column sidecar size, the data column indices and the SSZ encoded data column sidecar.
+	countToWrite := headerSize + len(sszEncodedDataColumnSidecars)
+	bytes := make([]byte, 0, countToWrite)
+	bytes = append(bytes, byte(version))
+	bytes = append(bytes, encodedSszEncodedDataColumnSidecarSize[:]...)
+	bytes = append(bytes, indices[:]...)
+	bytes = append(bytes, sszEncodedDataColumnSidecars...)
+
+	countWritten, err := file.Write(bytes)
+	if err != nil {
+		return errors.Wrap(err, "write")
+	}
```

Comment:
IMO we'll start switch to fsync by default for blobs before peerdas, so in this branch I'd just add `file.Sync` right around here.

Reply:
I don't see what adding `file.Sync` will bring here, since a file lines after we hit the `defer` statement containing the `file.Close` call. And closing the file dumps data into the disk right?

Reply:
`file.Sync` is `fsync` under the hood: https://linux.die.net/man/2/fsync

Calling close does not call fsync - data can still be in a cache waiting to be fully committed as written by the disk controller. Without calling fsync you may wind up with incomplete writes if there is a hard restart (which could happen for instance if the k8s pod is rescheduled or OOMKilled). This can result in corruption in the data file, extra important to avoid here where we have a header that needs to be consistent with the entries.

Reply:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/60db9a4d763d6267081e99d5659356ba4c011350.